### PR TITLE
buildPerlPackage: phase out "name"

### PR DESCRIPTION
**Pull request by volth**
Some time ago `perlPackages.*` have been refactored to use `buildPerlPackage` with `pname` and `version` instead of `name`
and a warning was printed when it was used with `name` to prompt users fix their code.
As time passed, now the warning is being replaced with a fatal error and the support for `name` is removed.

PS. rebuild is not expected

### CODE REVIEW

File: pkgs/development/perl-modules/generic/default.nix [hunk 1]
```diff
@@ -34,17 +31,17 @@ toPerlModule(stdenv.mkDerivation (
...
   )
   //
   {
-    pname = "perl${perl.version}-${attrs.pname or (builtins.parseDrvName attrs.name).name}"; # TODO: phase-out `attrs.name`
-    version = attrs.version or (builtins.parseDrvName attrs.name).version;                   # TODO: phase-out `attrs.name`
+    pname = "perl${perl.version}-${pname}";
+    inherit version;
     builder = ./builder.sh;
-    buildInputs = buildInputs ++ [ perl ];
+    buildInputs = buildInputs ++ [ perl ]; # the only reason for host platform's `perl` here is to support authomatic `patchShebang` of programs which may be in `$out/bin`
```
Comment (aanderse) [inline pkgs/development/perl-modules/generic/default.nix]: `authomatic` -> `automatic`

**Comment by worldofpeace** [comment]
I believe there was throw in `stdenv.mkDerivation` that did the same thing and it was reverted.
I think having the builder functions inconsistent like this would be confusing to users, for example every `buildPythonPackage` pretty much uses `pname` but it doesn't `throw` for `name`.
I think the warning is good if it only affects external expressions.

**Comment by volth** [comment]
> I believe there was throw in `stdenv.mkDerivation` that did the same thing and it was reverted.

I think `stdenv.mkDerivation` should print a warning right now (and maybe `throw` next year).

The `name` -> `pname` migration is stalled, up to new PRs get with `name` (well, I make them too, in the absence of warning it is easy to forget after years of using `name`)

The migration is important for various `lang2nix` and `repology` (https://github.com/NixOS/nixpkgs/issues/61691) because `name` is tricky to split to package name and version.

**Comment by worldofpeace** [comment]
> > I believe there was throw in `stdenv.mkDerivation` that did the same thing and it was reverted.
> 
> I think `stdenv.mkDerivation` should print a warning right now (and maybe `throw` next year).
> 
> The `name` -> `pname` migration is stalled, up to new PRs get with `name` (well, I make them too, in the absence of warning it is easy to forget after years of using `name`)
> 
> The migration is important for various `lang2nix` and `repology` (#61691) because `name` is tricky to split to package name and version.

Indeed all of this is true, and we'll clearly benefit from having `pname` and `version` everywhere.
So perhaps just a warning in `stdenv.mkDerivation` should be discussed?

**Comment by volth** [comment]
> So perhaps just a warning in `stdenv.mkDerivation` should be discussed?

Yes.
Replacing warning with throw is to be postponed till `20.03` or `20.09`

**Comment by volth** [comment]
Oh, a warning in `stdenv.mkDerivation` emits a lot of false positives
(so I doubt there has been a PR proposing `throw` in `stdenv.mkDerivation`)
```
warning: stdenv.mkDerivation:                       
  name = "d851fada460d42508a6f82b19867f63853062583.patch"  
is deprecated, please use                                  
  pname = "d851fada460d42508a6f82b19867f63853062583.patch";
  version = "";                                            
                                                           
warning: stdenv.mkDerivation:                       
  name = "source"                                          
is deprecated, please use                                  
  pname = "source";                                        
  version = "";                                            

warning: stdenv.mkDerivation:
  name = "daemon.conf"              
is deprecated, please use           
  pname = "daemon.conf";            
  version = "";                     

warning: stdenv.mkDerivation:
  name = "Net-DBus-1.1.0.tar.gz"    
is deprecated, please use           
  pname = "Net-DBus";               
  version = "1.1.0.tar.gz";         

warning: stdenv.mkDerivation:
  name = "unit-user-.service"       
is deprecated, please use           
  pname = "unit-user";              
  version = ".service";             

warning: stdenv.mkDerivation:                                                  
  name = "CVE-2017-6827+CVE-2017-6828+CVE-2017-6832+CVE-2017-6835+CVE-2017-6837.patch"
is deprecated, please use                                                             
  pname = "CVE";                                                                      
  version = "2017-6827+CVE-2017-6828+CVE-2017-6832+CVE-2017-6835+CVE-2017-6837.patch";

warning: stdenv.mkDerivation:      
  name = "ntfs-3g_ntfsprogs-2017.3.23.tgz"
is deprecated, please use                 
  pname = "ntfs";                         
  version = "3g_ntfsprogs-2017.3.23.tgz"; 

```

**Comment by worldofpeace** [comment]
Ah it was an assert that was reverted from the original implementation #51183

**Comment by volth** [comment]
BTW, some language builders do not support `name`.

https://github.com/NixOS/nixpkgs/blob/ad135f051837538b7486250075e94540018feb2d/pkgs/build-support/emacs/generic.nix#L7-L8
https://github.com/NixOS/nixpkgs/blob/ad135f051837538b7486250075e94540018feb2d/pkgs/build-support/build-dotnet-package/default.nix#L4-L5
https://github.com/NixOS/nixpkgs/blob/ad135f051837538b7486250075e94540018feb2d/pkgs/build-support/ocaml/oasis.nix#L3

So perl won't be pioneering if `pname` and `version` will be required params of `buildPerlPackage`

**Comment by worldofpeace** [comment]
I'd say those are bugs that should be fixed before release.

**Comment by worldofpeace** [comment]
If we're going to do this we'd need to update all the examples at
* https://nixos.org/nixpkgs/manual/#sec-language-perl

and still people may think they could use `name` but at least it will give a message.

**Comment by volth** [comment]
Let's keep it open and collect people opinions

**Comment by volth** [comment]
On the other hand, `fetchurl` (which is kind of default builder for `data/fonts/*`) does not accept `pname`
That result in weird code
https://github.com/NixOS/nixpkgs/blob/94b155aa8be0b039c5cf5bc3b95771609f2af401/pkgs/data/fonts/agave/default.nix#L3-L7

**Pull request closed without merging**